keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1004 @@
+particular TrustAnchorSets using a trustAnchorSets volume?  The straightforward
+design would have all TrustAnchorSet reads happen using a `system:node:xyz`
+identity, which would allow any workload to mount any TrustAnchorSets object.
+
+Security: Should individual trust anchor set entries designate an OSCP endpoint
+to check for certificate revociation?  Or should we require the URL to be
+embedded in the issued certificates?
+
+Reliability: How can we safely canary a change to a TrustAnchorSets object?
+
+Usability: Is it always necessary to provide a value for spec.signerName?
+
+## Design Details
+
+### TrustAnchorSets Object
+
+The TrustAnchorSets object allows signers to publish their trust anchors for
+consumption by workloads, whether directly, via Kubelet, or otherwise.
+
+Every signer corresponds to zero, one, or more TrustAnchorSets objects.  The
+name of the TrustAnchorSets objects are arbitrary; the association is managed
+entirely via the `spec.signerName` field, which is immutable.
+
+Since a given signer could have multiple TrustAnchorSets objects, clients
+looking for a particular named set X must take the union of all `spec.sets`
+entries where `spec.sets[i].name == X`.
+
+#### API Object Definition
+
+```go
+// TrustAnchorSets packages up trust anchors.
+//
+// Each TrustAnchorSets object is owned by a particular signer, which
+// updates the trust anchors as the roots of trust for that signer change.
+//
+// A TrustAnchorSets object wraps up multiple trust anchor sets, each of
+// which in turn is a set of one or more trust anchors (individual
+// certificates).  Most use cases will only require a single trust
+// anchor set, conventionally named `default`.  Multiple uniquely-named trust
+// anchor sets will likely only be required in mutual TLS use cases with
+// federation between multiple meshes.
+type TrustAnchorSets struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec TrustAnchorSetsSpec `json:"spec"`
+  Status TrustAnchorSetsStatus `json:"status"`
+}
+
+type TrustAnchorSetsSpec struct {
+  // The name of the signer that owns this TrustAnchorSets object.
+  //
+  // Each signer may maintain multiple TrustAnchorSets objects.
+  SignerName string
+}
+
+type TrustAnchorSetsStatus struct {
+  // The list of trust anchor sets.
+  Sets []TrustAnchorSet
+
+  // The current conditions of the TrustAnchorSets object.
+  //
+  // There is one canonical condition, "Ready", that the signer uses
+  // to indicate to kubelet that the TrustAnchorSets object is up to date.
+  // This is used to keep pods from starting up with empty trust anchor sets
+  // while the signer is initializing.
+  Conditions []metav1.Condition
+}
+
+type TrustAnchorSet struct {
+  // The unique name of this trust anchor set.
+  Name string
+
+  // The individual trust anchors for this set.  Each entry is a single
+  // PEM-wrapped, DER-formatted X.509 certificate.
+  PEMTrustAnchors []string
+}
```

Comment:
The go std lib has a `ParseCRL` function which has me thinking.

Could we have an optional `PEMCRLs []string` function that allows the trust anchor set to specify a set of CRLs files that are synced on to the file system in the same way as the CA bundle?  It would be up to the client to read and honor them of course.

Reply:
had a quick read of the KEP and i don't think i understand all the implications, but having CRL here makes sense to me.

Reply:
I don't think we should be embedding a CRL directly within the Kubernetes object, for the following reasons:

* It seems like CRLs have the potential to grow very large, especially if you are using moderately long-lived certificates and large deployments.  It wouldn't be great if you could only revoke 50% of the certificates you need to revoke because of etcd object size limits.
* This would be custom to Kubernetes.  Users would have to write integration code to load the CRL from the filesystem and plumb it into their TLS stack.  If users are going to have to write custom code, they might was well write custom code that loads an OCSP url from an X.509 extension in the root certificate.

It might be workable to put an OCSP URL in the Kubernetes object, but again I'm not sure what the value-add is over just putting it in the root certificate.

Reply:
I'd prefer that we recommend that people put the CRL distribution point and OCSP url in the certificate of the CA.